File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/controller/WebViewReportController.java
/*
 * KurobaEx - *chan browser https://github.com/K1rakishou/Kuroba-Experimental/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.k1rakishou.chan.ui.controller;

import static com.github.k1rakishou.chan.utils.AppModuleAndroidUtils.getString;
import static com.github.k1rakishou.chan.utils.AppModuleAndroidUtils.inflate;

import android.annotation.SuppressLint;
import android.content.Context;
import android.util.AndroidRuntimeException;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.github.k1rakishou.chan.R;
import com.github.k1rakishou.chan.controller.Controller;
import com.github.k1rakishou.chan.core.di.component.activity.ActivityComponent;
import com.github.k1rakishou.chan.core.navigation.RequiresNoBottomNavBar;
import com.github.k1rakishou.chan.core.site.Site;
import com.github.k1rakishou.chan.core.site.SiteRequestModifier;
import com.github.k1rakishou.common.KotlinExtensionsKt;
import com.github.k1rakishou.model.data.post.ChanPost;
import com.github.k1rakishou.model.util.ChanPostUtils;

import org.jetbrains.annotations.NotNull;

import okhttp3.HttpUrl;

public class WebViewReportController extends Controller implements RequiresNoBottomNavBar {
    private ChanPost post;
    private Site site;
    private int toolbarHeight = 0;

    @Override
    protected void injectDependencies(@NotNull ActivityComponent component) {
        component.inject(this);
    }

    public WebViewReportController(Context context, ChanPost post, Site site, int toolbarHeight) {
        super(context);
        this.post = post;
        this.site = site;
        this.toolbarHeight = toolbarHeight;
    }

    @SuppressLint("SetJavaScriptEnabled")
    @Override
    public void onCreate() {
        super.onCreate();
        navigation.title = getString(R.string.report_screen, ChanPostUtils.getTitle(post, null));

        HttpUrl url = site.endpoints().report(post);

        FrameLayout frameLayout = new FrameLayout(context);
        frameLayout.setLayoutParams(
                new LinearLayout.LayoutParams(
                        FrameLayout.LayoutParams.MATCH_PARENT,
                        FrameLayout.LayoutParams.MATCH_PARENT
                )
        );

        try {
            WebView webView = new WebView(context);

            SiteRequestModifier<Site> siteRequestModifier = site.requestModifier();
            if (siteRequestModifier != null) {
                siteRequestModifier.modifyWebView(webView);
            }

            WebSettings settings = webView.getSettings();
            settings.setJavaScriptEnabled(true);
            settings.setDomStorageEnabled(true);
            webView.loadUrl(url.toString());

            frameLayout.addView(
                    webView,
                    new FrameLayout.LayoutParams(
                            FrameLayout.LayoutParams.MATCH_PARENT,
                            FrameLayout.LayoutParams.MATCH_PARENT
                    )
            );

            view = frameLayout;

            KotlinExtensionsKt.updatePaddings(
                    frameLayout,
                    null,
                    null,
                    toolbarHeight,
                    null
            );

        } catch (Throwable error) {
            String errmsg = "";
            if (error instanceof AndroidRuntimeException && error.getMessage() != null) {
                if (error.getMessage().contains("MissingWebViewPackageException")) {
                    errmsg = getString(R.string.fail_reason_webview_is_not_installed);
                }
            } else {
                errmsg = getString(R.string.fail_reason_some_part_of_webview_not_initialized, error.getMessage());
            }
            view = inflate(context, R.layout.layout_webview_error);
            ((TextView) view.findViewById(R.id.text)).setText(errmsg);
        }
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/controller/WebViewReportController.kt
package com.github.k1rakishou.chan.ui.controller

import android.annotation.SuppressLint
import android.content.Context
import android.util.AndroidRuntimeException
import android.webkit.CookieManager
import android.webkit.WebView
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import com.github.k1rakishou.chan.R
import com.github.k1rakishou.chan.core.di.component.activity.ActivityComponent
import com.github.k1rakishou.chan.core.manager.GlobalWindowInsetsManager
import com.github.k1rakishou.chan.core.manager.WindowInsetsListener
import com.github.k1rakishou.chan.core.site.Site
import com.github.k1rakishou.chan.features.toolbar.BackArrowMenuItem
import com.github.k1rakishou.chan.features.toolbar.ToolbarMiddleContent
import com.github.k1rakishou.chan.features.toolbar.ToolbarText
import com.github.k1rakishou.chan.ui.controller.base.Controller
import com.github.k1rakishou.chan.utils.AppModuleAndroidUtils
import com.github.k1rakishou.common.CookieBuilder
import com.github.k1rakishou.common.resumeValueSafe
import com.github.k1rakishou.common.updatePaddings
import com.github.k1rakishou.core_logger.Logger
import com.github.k1rakishou.model.data.post.ChanPost
import com.github.k1rakishou.model.util.ChanPostUtils.getTitle
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import javax.inject.Inject

class WebViewReportController(
  context: Context,
  private val post: ChanPost,
  private val site: Site
) : Controller(context), WindowInsetsListener {

  @Inject
  lateinit var globalWindowInsetsManager: GlobalWindowInsetsManager

  private val cookieManager by lazy { CookieManager.getInstance() }

  private lateinit var frameLayout: FrameLayout

  override fun injectActivityDependencies(component: ActivityComponent) {
    component.inject(this)
  }

  @SuppressLint("SetJavaScriptEnabled")
  override fun onCreate() {
    super.onCreate()

    toolbarState.enterDefaultMode(
      leftItem = BackArrowMenuItem(
        onClick = { requireNavController().popController() }
      ),
      middleContent = ToolbarMiddleContent.Title(
        title = ToolbarText.String(appResources.string(R.string.report_screen, getTitle(post, null)))
      )
    )

    controllerScope.launch {
      try {
        // Some users may have no WebView installed
        initUi()
      } catch (error: Throwable) {
        Logger.e(TAG, "Error when trying to create the view", error)
        requireNavController().popController()
        return@launch
      }

      controllerScope.launch {
        combine(
          globalUiStateHolder.toolbar.toolbarHeight,
          globalUiStateHolder.bottomPanel.bottomPanelHeight
        ) { t1, t2 -> t1 to t2 }
          .onEach { onInsetsChanged() }
          .collect()
      }

      onInsetsChanged()
      globalWindowInsetsManager.addInsetsUpdatesListener(this@WebViewReportController)
    }
  }

  override fun onDestroy() {
    super.onDestroy()

    globalWindowInsetsManager.removeInsetsUpdatesListener(this)
  }

  override fun onInsetsChanged() {
    val bottomPadding = with(appResources.composeDensity) {
      maxOf(
        globalWindowInsetsManager.bottom(),
        globalUiStateHolder.bottomPanel.bottomPanelHeight.value.roundToPx()
      )
    }

    val topPadding = with(appResources.composeDensity) {
      maxOf(
        globalWindowInsetsManager.top(),
        globalUiStateHolder.toolbar.toolbarHeight.value.roundToPx()
      )
    }

    frameLayout.updatePaddings(
      top = topPadding,
      bottom = bottomPadding
    )
  }

  private suspend fun initUi() {
    val urlToOpen = site.endpoints().report(post)
    if (urlToOpen == null) {
      requireNavController().popController()
      return
    }

    frameLayout = FrameLayout(context)
    frameLayout.setLayoutParams(
      LinearLayout.LayoutParams(
        FrameLayout.LayoutParams.MATCH_PARENT,
        FrameLayout.LayoutParams.MATCH_PARENT
      )
    )

    try {
      val webView = WebView(context)
      val siteRequestModifier = site.requestModifier()

      suspendCancellableCoroutine { cont ->
        cookieManager.removeAllCookies {
          Logger.debug(TAG) { "cookieManager.removeAllCookies()" }
          cont.resumeValueSafe(Unit)
        }
      }

      if (siteRequestModifier != null) {
        val cookieManager = CookieManager.getInstance()
        val cookieBuilder = CookieBuilder()
        siteRequestModifier.modifyCookieBuilder(urlToOpen, cookieBuilder)

        val builtCookies = cookieBuilder.build()
        if (builtCookies.isNotBlank()) {
          cookieManager.setCookie(urlToOpen.toString(), builtCookies)
        }
      }

      val settings = webView.getSettings()
      settings.javaScriptEnabled = true
      settings.domStorageEnabled = true
      webView.loadUrl(urlToOpen.toString())

      frameLayout.addView(
        webView,
        FrameLayout.LayoutParams(
          FrameLayout.LayoutParams.MATCH_PARENT,
          FrameLayout.LayoutParams.MATCH_PARENT
        )
      )

      view = frameLayout
      onInsetsChanged()
    } catch (error: Throwable) {
      var errmsg = ""

      if (
        error is AndroidRuntimeException &&
        error.message != null &&
        error.message?.contains("MissingWebViewPackageException") == true
      ) {
        errmsg = appResources.string(R.string.fail_reason_webview_is_not_installed)
      } else {
        errmsg = appResources.string(
          R.string.fail_reason_some_part_of_webview_not_initialized,
          error.message ?: "Unknown error"
        )
      }

      view = AppModuleAndroidUtils.inflate(context, R.layout.layout_webview_error)
      view.findViewById<TextView>(R.id.text).text = errmsg
    }
  }

  companion object {
    private const val TAG = "WebViewReportController"
  }

}
